demo/src/main/java/com/smashingboxes/surelockdemo/LoginActivity.kt

```diff
@@ -0,0 +1,346 @@
+        } else {
+            attemptLogin(username, password)
+        }
+    }
+
+    private fun attemptLogin(usernameInput: String?, passwordInput: String?) {
+        // Don't allow multiple login attempts in a row
+        if (authTask != null) {
+            return
+        }
+
+        if (!validateLoginForm(usernameInput, passwordInput)) {
+            showProgress(true)
+
+            authTask = UserLoginTask(false)
+            //We're just fetching this to show you on the next screen for demo purposes.
+            var storedEncryptedValueString = ""
+            try {
+                val storedEncryptedValue = surelockStorage.get(KEY_CRE_DEN_TIALS)
+                if (storedEncryptedValue != null) {
+                    storedEncryptedValueString = String(storedEncryptedValue, Charsets.UTF_8)
+                }
+            } catch (e: UnsupportedEncodingException) {
+                Log.d(TAG, "attemptLogin: cannot fetch encrypted credentials")
+            }
+
+            authTask!!.execute(
+                username.text.toString(), password.text.toString(),
+                storedEncryptedValueString,
+                usernameInput, passwordInput)
+        }
+    }
+
+    private fun attemptLoginForFingerprintEnrollment(usernameInput: String, passwordInput: String) {
+        // Don't allow multiple login attempts in a row
+        if (authTask != null) {
+            return
+        }
+
+        if (!validateLoginForm(usernameInput, passwordInput)) {
+            showProgress(true)
+            authTask = UserLoginTask(true)
+            authTask!!.execute(username.text.toString(), password.text.toString())
+        }
+    }
+
+
+    private fun showProgress(show: Boolean) {
+        val shortAnimTime = resources.getInteger(android.R.integer.config_shortAnimTime)
+
+        login_form.visibility = if (show) View.GONE else View.VISIBLE
+        login_form.animate()
+            .setDuration(shortAnimTime.toLong())
+            .alpha((if (show) 0 else 1).toFloat())
+            .setListener(object : AnimatorListenerAdapter() {
+                override fun onAnimationEnd(animation: Animator) {
+                    login_form.visibility = if (show) View.GONE else View.VISIBLE
+                }
+            })
+
+        login_progress.visibility = if (show) View.VISIBLE else View.GONE
+        login_progress.animate()
+            .setDuration(shortAnimTime.toLong())
+            .alpha((if (show) 1 else 0).toFloat())
+            .setListener(object : AnimatorListenerAdapter() {
+                override fun onAnimationEnd(animation: Animator) {
+                    login_progress.visibility = if (show) View.VISIBLE else View.GONE
+                }
+            })
+    }
+
+    /**
+     * Represents an asynchronous login/registration task
+     * used to authenticate the user.
+     */
+    private inner class UserLoginTask(private val withFingerprintEnrollment: Boolean) :
```

Comment:
I think you're not supposed to have AsyncTasks as inner classes.
They can cause memory leaks if they maintain a reference to the view.
I would at the very least pass a reference to the fragment/activity/view in the constructor and set it as a `WeakReference<>` type and also cancel the AsyncTask in `onStop()`.

Reply:
Since this is just a demo module. You're not going to hurt anyone's feelings if you just leave it alone though lol